core/src/main/scala/io/snappydata/SnappyTableStatsProviderService.scala

```diff
@@ -207,19 +210,36 @@ object SnappyEmbeddedTableStatsProviderService extends TableStatsProviderService
       }
     }
     catch {
-      case NonFatal(e) => log.warn(e.getMessage, e)
+      case NonFatal(e) => {
+        log.warn("Exception occurred while collecting External Table Statistics: "
+            + e.getMessage, e)
+      }
     }
 
-    if(result.flatMap(_.getRegionStats.asScala).size == 0) {
-      // Return last updated tableSizeInfo
+    if (result.flatMap(_.getRegionStats.asScala).size == 0) {
+      // Return last successfully updated tableSizeInfo
+      // Check for if any table is present in catalog
+      val allStoreTables = Misc.getMemStore.getExternalCatalog.getAllStoreTablesInCatalog(true)
+      val allTablesList = scala.collection.mutable.Buffer.empty[String]
+      allStoreTables.asScala.keys.foreach(schema => {
+        allTablesList ++= allStoreTables.get(schema).asScala.map(tbl => {
+          schema.concat(".").concat(tbl)
+        })
+      })
+
```

Comment:
Why store in a list when you only need to know whether size is 0 or not? Also shouldn't the externalTables and indexStats also be cleared.

This size==0 does not look to be proper. I think you should simply use a flag to indicate the case when result was successfully acquired:

```scala
      if (dataServers != null && dataServers.size() > 0) {
        result = FunctionService.onMembers(dataServers)
            // .withCollector(new GfxdListResultCollector())
            .execute(SnappyRegionStatsCollectorFunction.ID).getResult(5, TimeUnit.SECONDS).
            asInstanceOf[java.util.ArrayList[SnappyRegionStatsCollectorResult]]
            .asScala
        resultObtained = true
      }
```

And use that flag instead of size check. Then there is no need to go back to catalog again so this new code can be removed.

Reply:
Feedback is incorporated in the code.